File: adventofcode/2025/src/day3.ts
import { generateMain, getInputStrings } from "./framework";

// Puzzle config.
const DAY = 3;
const SAMPLE_ANSWER_1 = 357;
const SAMPLE_ANSWER_2 = 3121910778619;

// Types.
type BatteryBank = number[];

type Input = BatteryBank[];


// Parse input.
const parseInput = (input: string): Input =>
    input.split('\n').map((line) => line.split('').map(Number));


// Solve.
const solve1 = (input: Input): number => {
    return input.reduce((prev, curr) => prev + findMaxJoltageOfBatteryBank(curr, 2), 0);
};

const solve2 = (input: Input): number => {
    return input.reduce((prev, curr) => prev + findMaxJoltageOfBatteryBank(curr, 12), 0);
};

function findMaxJoltageOfBatteryBank(bank: BatteryBank, length: number): number {
    if (!bank.length) {
        return 0;
    }

    if (bank.length <= length) {
        return Number(bank.join(''));
    }

    return Number(findSequence(bank, length, []).join(''));
}

function findSequence(bank: BatteryBank, length: number, sequence: BatteryBank): BatteryBank {
    if (sequence.length >= length) {
        return sequence;
    }

    const remainingLength = length - sequence.length;

    const order = [9, 8, 7, 6, 5, 4, 3, 2, 1, 0];
    for (const digit of order) {
        const index = bank.indexOf(digit);
        if (index === -1 || index + remainingLength > bank.length) {
            continue;
        }

        return findSequence(bank.slice(index + 1), length, [...sequence, digit]);
    }

    throw new Error('Cannot find any digit?!');
}


// Main.
generateMain(parseInput, solve1, solve2)(...(await getInputStrings(DAY)), SAMPLE_ANSWER_1, SAMPLE_ANSWER_2);

File: adventofcode/2024/src/day3.py
import re
from functools import reduce

from framework import run_solutions

# Configuration
DAY = 3
SAMPLE_ANSWER1 = 161
SAMPLE_ANSWER2 = 48


# Solutions
mul_regex = r'mul\((\d{1,3}),(\d{1,3})\)'

def solve1(input: str) -> int:
    matches = re.finditer(mul_regex, input)
    return reduce(lambda prev, curr: prev + int(curr[1]) * int(curr[2]), matches, 0)


def solve2(input: str) -> int:
    matches = re.finditer(r"(do(?:n't)?\(\))|(mul\(\d{1,3},\d{1,3}\))", input)
    total = 0
    is_enabled = True
    for match in matches:
        if match[0] == 'do()':
            is_enabled = True
        elif match[0] == "don't()":
            is_enabled = False
        elif is_enabled:
            digits = re.findall(mul_regex, match[0])[0]
            total += int(digits[0]) * int(digits[1])

    return total


# Run solutions
run_solutions(DAY, (solve1, solve2), (SAMPLE_ANSWER1, SAMPLE_ANSWER2))
